docs/rfc/0003-collaborative-objects.md

```diff
@@ -0,0 +1,421 @@
+layer.
+
+## Design Goals
+
+- Local first
+- Extensible. It should be possible to extend Radicle with new collaborative
+  data types without changing the protocol version
+- Interoperable. There should be a straightforward API for disparate tools to
+  interact with collaborative objects
+
+Local first because you should not need to be online to modify collaborative
+objects and no one should be able to stop you from changing data that is on
+your computer. That said, collaborative objects are _collaborative_, users need
+to be able to receive changes to the same data from other people and this
+raises the problem of how to merge those changes. I see two options here:
+
+- CRDTs, data structures which allow conflict free merges
+- Application level merges. Much as Git requires user action to merge
+  conflicting changes, we could expose the merge points of collaborative
+  objects to the user to resolve.
+
+The latter here is undesirable for a lot of reasons but most relevant is that
+it conflicts with extensibility and interoperability. in order to provide a good 
+UX tools would need to provide a UI for users to do a three way merge of the 
+data they are working on. This contradicts the interoperability goal.
+
+This leaves us with CRDTs. For the purposes of this RFC I am going to assume
+that we will use the [Automerge](https://github.com/automerge/automerge) CRDT
+implementation. Automerge is capable of merging arbitrary JSON objects (the
+data model is actually richer than JSON, with types for byte arrays, various
+precision integers and floats etc.). An alternative approach would be to write
+a custom CRDT for each data type we want to replicate, see [##Alternative
+Approaches] for a discussion of this design.
+
+## Example: Issues
+
+To motivate and contextualise this proposal we introduce a running example
+which is dear to many peoples hearts; issues. We imagine this from the
+perspective of the application developer. For expositional purposes we assume
+that the application communicates with a radicle implementation via an HTTP
+RPC. However, the HTTP RPC is not a proposal of this RFC, the proposed API will
+be specified in terms of `librad` later in this document.
+
+The first thing the developer must do is decide on the schema of their data and
+represent it as a JSON schema. We use this simple schema:
+
+```json
+{
+  "type": "object"
+  "properties": {
+    "title": {
+        "type": "string"
+    },
+    "description": {
+        "type": "string"
+    },
+    "author": {
+        "type": "string",
+        "description": "The radicle ID of the author of the issue"
+    },
+    "signature": {
+        "type": "string",
+        "description": "A base64 encoded signature of the issue"
+    },
+    "comments": {
+        "type": "array",
+        "items": {
+            "type": "object",
+            "properties": {
+                "text": {"type": "string"},
+                "author": {
```

Comment:
I'm curious if there'll be ways to define common components, for example `author` is repeated here. Or will we have to reach for something like Dhall to do that? :)